### PR TITLE
Issues with resource quotas

### LINKED ISSUE #4740: Issues with Resource Quotas
**What is the problem, where is it and why is it an issue?**

1. Alignment issues (*isolated 😁)

Screenshot 2022-07-30 at 9 55 01 AM

2. We should use a typeahead here and project names should be sorted. It was very difficult for me to find my project.

Screenshot 2022-07-30 at 9 45 57 AM (2)

3. We shouldn't send in patch requests when nothing changed. Check if the form/controls changed, `this.control(XYZ).pristine`. 

https://user-images.githubusercontent.com/18264334/181875746-a9ee16ca-e831-4e43-a604-2784a6643248.mov

4. There is something extremely wrong with these input fields. It just keeps flushing the input if it exceeds 3 characters in case of `Memory` and `Disk`. For CPU if I enter `e` then it just removes the whole input, which is wrong. In general either don't handle `e` or ignore it.

https://user-images.githubusercontent.com/18264334/181875853-8516801c-c9a5-4ff2-a6f0-58baeb7e3dc6.mov

5. Maybe I'm not quite understanding this but the quota limitations didn't seem to work for me. This has nothing to do with the front-end IMO. I set a resource quota for 1 CPU and chose a flavor on AWS that has 2 vCPUs. It didn't restrict me in any way and I got a working cluster.

Screenshot 2022-07-30 at 10 05 18 AM
Screenshot 2022-07-30 at 10 05 35 AM
Screenshot 2022-07-30 at 10 08 21 AM

6. We should show units here in the widget.

Screenshot 2022-08-02 at 5 18 36 PM

7. Why are we polling for quotas so so often? This is from the cluster details view.

Screenshot 2022-08-02 at 10 07 09 PM
Screenshot 2022-08-02 at 10 05 20 PM

### PR DESCRIPTION
**Pull request by Talha-Jamil-TJ**
**What this PR does / why we need it**:

### Point 1 (b) & 3

While in `edit mode`:
- Project name is moved to the top
- submit button will stay disabled when quotas are unchanged
image

### Point 2

Select component is replaced with a typeahead 

image

### Point 4

Issues with number inputs are fixed

### Point 6

Units added to quota widget on hover

image

### Point 8

Fixed background colour of yellow progress bar

image
image

**Which issue(s) this PR fixes**:

Fixes #4740

**What type of PR is this?**
/kind bug
/kind design

**Documentation**:

```documentation
NONE
```

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
NONE
```

### CODE REVIEW

File: src/app/cluster/details/cluster/edit-cluster/template.html
Comment (Waseem826) [inline src/app/cluster/details/cluster/edit-cluster/template.html:99]: This has already been fixed in https://github.com/kubermatic/dashboard/pull/4800 so you can revert this.

**Comment by Waseem826** [comment]
With this updated Project field, can we now enter custom Project name? If not, then why can't we use `km-combobox` which already has search field for filtering? And it also supports custom markup for options.

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/template.html
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/template.html:95]:
```suggestion
              <mat-card-subtitle class="km-no-margin km-no-padding"
```
So that it is aligned with project name.

File: src/app/shared/components/number-stepper/component.ts
Comment (Waseem826) [inline src/app/shared/components/number-stepper/component.ts:121]: What was the case which required to add `replace(/,/g, '')`?

File: src/app/shared/components/number-stepper/template.html
Comment (Waseem826) [inline src/app/shared/components/number-stepper/template.html:24]: We shouldn't be ignore `e` in this component since its a generic component but instead this should be handled in the parent component where we want to ignore `e`

File: src/app/cluster/details/cluster/edit-cluster/template.html
Comment (Talha-Jamil-TJ) [reply src/app/cluster/details/cluster/edit-cluster/template.html:99]: This is happening when I run `npm run fix`, if I don't run it, pipeline will fail.

Comment (Waseem826) [reply src/app/cluster/details/cluster/edit-cluster/template.html:99]: Pipeline won't fail because of this particular case.

File: src/app/dynamic/enterprise/quotas/quota-widget/template.html
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/quota-widget/template.html:72]: this part is still being updated, Once I have approval from @cschieder then we can discuss it further.

File: src/app/shared/components/number-stepper/component.ts
Comment (Talha-Jamil-TJ) [reply src/app/shared/components/number-stepper/component.ts:121]: once we enter upto four digits `ex: 1111`, the method return `1,111`, which not supported by number input and cannot be parsed too, that's why I'm removing the `,` so the value can be parsed.

File: src/app/shared/components/number-stepper/template.html
Comment (Talha-Jamil-TJ) [reply src/app/shared/components/number-stepper/template.html:24]: when we enter `e` in the text input it clears the input, this issue is present everywhere this component is being used.
This was done after discussion with @ahmedwaleedmalik

File: src/app/shared/components/number-stepper/component.ts
Comment (Waseem826) [reply src/app/shared/components/number-stepper/component.ts:121]: But this is only for `decimal` type and we're not using that in Resource quotas. Also, `1.0-4` should not return value with `,`.
In the screenshot below I entered 9 digits without any issue (on dev)

screenshot-localhost_8000-2022 08 16-14_47_58

This is what Angular docs say about the `digitsInfo` parameter: https://angular.io/api/common/DecimalPipe#digitsinfo

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:137]: Instead of running `filter` for every project (hence increasing the complexity), can we do the following?
```suggestion
      this.projectNameCountMap = {};
      projects.forEach(project => {
        this.projectNameCountMap[project.id] = (this.projectNameCountMap[project.id] || 0) + 1;
      });
```

File: src/app/cluster/details/cluster/edit-cluster/template.html
Comment (Waseem826) [inline src/app/cluster/details/cluster/edit-cluster/template.html:99]: This has been fixed in `master` so please revert this change and rebase your branch on latest `master`.

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:60]: I think there's no need for this comment and the one below because property names are self explanatory.

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:99]: Why can't we use `projectIdAndNameMap?.[projectId]` directly in template instead of calling this method?

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:148]: Please move this statement inside a block

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:183]: I think we don't need this validator anymore since we've now switched to `km-combobox`

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Waseem826) [inline src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:188]: Why can't we use `subjectName` control instead of this?

**Review by Waseem826 — CHANGES_REQUESTED** [review]

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:188]: The `km-combobox` component returns a value which is nested inside an object (`{ select: <actual data> }`), which is not representative of the payload I need to send to the API, this is why I'm handling it in a separate formControl.

File: src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts
Comment (Talha-Jamil-TJ) [reply src/app/dynamic/enterprise/quotas/project-quota-dialog/component.ts:96]: we need a function to pass as input to km-combobox

**Review by Waseem826 — APPROVED** [review]
/approve

**Pull request merged**
